ffi_internals/src/impl_internals/fn_ffi.rs

```diff
@@ -172,94 +339,43 @@ impl FnFFI {
         }
     }
 
-    fn ffi_fn_name(&self, module_name: &Ident) -> Ident {
+    pub(crate) fn ffi_fn_name(&self, module_name: &Ident) -> Ident {
         format_ident!("{}_{}", module_name, self.fn_name)
     }
-
-    /// Generates a consumer function for calling the foreign function produced by
-    /// `self.generate_ffi(...)`.
-    ///
-    pub fn generate_consumer(&self, module_name: &Ident) -> String {
-        let (return_conversion, close_conversion, return_sig) =
-            if let Some(return_type) = &self.return_type {
-                let ty = return_type.consumer_type(None);
-                (
-                    format!("{}.fromRust(", ty),
-                    format!(")"),
-                    format!("-> {}", ty),
-                )
-            } else {
-                (String::new(), String::new(), String::new())
-            };
-        format!(
-            r#"
-    func {native_fn_name}({native_parameters}) {return_sig} {{
-        {return_conversion}{ffi_fn_name}({ffi_parameters}){close_conversion}
-    }}
-            "#,
-            native_fn_name = self.fn_name.to_string(),
-            native_parameters = self.consumer_parameters(),
-            return_sig = return_sig,
-            return_conversion = return_conversion,
-            ffi_fn_name = self.ffi_fn_name(module_name).to_string(),
-            ffi_parameters = self.ffi_calling_arguments(),
-            close_conversion = close_conversion,
-        )
-    }
-
-    fn consumer_parameters(&self) -> String {
-        self.parameters
-            .iter()
-            .map(|arg| {
-                format!(
-                    "{}: {}",
-                    arg.name.to_string(),
-                    arg.native_type_data.consumer_type(None)
-                )
-            })
-            .collect::<Vec<String>>()
-            .join(", ")
-    }
-
-    fn ffi_calling_arguments(&self) -> String {
-        let mut parameters: Vec<String> = self
-            .parameters
-            .iter()
-            .map(|arg| format!("{}.toRust()", arg.name.to_string()))
-            .collect();
-        if self.has_receiver {
-            let receiver_arg = "pointer".to_string();
-            parameters.insert(0, receiver_arg);
-        }
-        parameters.join(", ")
-    }
 }
 
 /// Represents a parameter for to a Rust function.
 #[derive(Debug)]
-pub struct FnParameterFFI {
+pub(crate) struct FnParameterFFI {
     /// The name of this parameter.
     ///
-    pub name: Ident,
+    pub(crate) name: Ident,
 
     /// The type information for generating an FFI for this parameter.
     ///
-    pub native_type_data: NativeTypeData,
+    pub(crate) native_type_data: TypeFFI,
 
     /// The original type of the fn parameter.
     ///
-    pub original_type: Type,
+    pub(crate) original_type: Type,
 }
 
-impl From<&PatType> for FnParameterFFI {
-    fn from(arg: &PatType) -> Self {
+impl From<(&PatType, Vec<Ident>, Option<Ident>)> for FnParameterFFI {
```

Comment:
What's a Pat?

Reply:
![This is a pat](https://media1.giphy.com/media/uw3fTCTNMbXAk/giphy.gif?cid=790b76115bbbae4b8e5a503aa8367da1403c64ab11982dd9&rid=giphy.gif&ct=g)

It's a type ascription pattern https://docs.rs/syn/1.0.75/syn/struct.PatType.html -- so for the purposes of `fn_ffi`, a function parameter's identifier and type.

Reply:
Converted this to take a `FnParameterFFIInputs` with docs instead of the oversized tuple.